rfc/20241021-dynamic-provider-instances.md

```diff
@@ -0,0 +1,766 @@
+The above configuration means that `aws.foo` in the child module refers to the same provider instance as `aws.use1` refers to in the parent module.
+
+The mapping assigned to the `providers` argument is a static mapping, so it must always use the object constructor syntax with braces `{ }`; it's not valid to assign a dynamic expression that would generate a map.
+
+To build the dependency graph OpenTofu needs to bind each provider configuration address in the child module to exactly one provider configuration address in the parent module before evaluating any expressions. It's valid to use dynamic instance expressions on the right side of the mappings in the `providers` block, but as with resources all instances of a module call must refer to instances of the same provider configuration:
+
+```hcl
+provider "aws" {
+  alias    = "by_region"
+  for_each = var.aws_regions
+
+  region = each.key
+}
+
+module "example2" {
+  source = "./another-example" # This is _not_ the directory containing the earlier example
+  for_each  = var.aws_regions
+  providers = {
+    aws = aws.by_region[each.key]
+  }
+}
+```
+
+The above means that in each instance of `module.example2` the default instance of "aws" is dynamically decided by evaluating `each.key`. The resources using `provider = aws` (implicitly or explicitly) in that module all depend on the `aws.by_region` configuration from the parent module in the dependency graph, and then choose one of its instances dynamically at runtime after evaluating `each.key`.
+
+The final interesting case is passing a whole multi-instance provider configuration between modules. For this we extend the `configuration_aliases` syntax to allow declaring that the module expects to be passed a multi-instance configuration rather than a singleton configuration:
+
+```hcl
+variable "aws_regions" {
+  type = map(object({
+    cidr_block = string
+  }))
+}
+
+terraform {
+  required_providers {
+    aws = {
+      source = "hashicorp/aws"
+      configuration_aliases = [
+        aws.by_region[*],
+      ]
+    }
+  }
+}
+
+resource "aws_vpc" "main" {
+  for_each = var.aws_regions
+  provider = aws.by_region[each.key]
+
+  cidr_block = each.value.cidr_block
+  # ...
+}
+```
+
+When a module is configured in this way, the `providers` argument in any call _must_ include an entry for `aws.by_region` whose selected configuration also uses `for_each`, like this:
+
+```hcl
+provider "aws" {
+  alias    = "by_region"
+  for_each = var.aws_regions
+
+  region = each.key
+}
+
+module "example" {
+  source = "./example" # This is the directory containing the previous example block
+  providers = {
+    aws.by_region = aws.by_region
+  }
+
+  # ...
+}
+```
+
+In _this_ situation `aws.by_region` in both the parent and child modules refers to the same multi-instance provider configuration, with the same instance keys. The child module allows the parent module to decide which instance keys to use, so the child module can be written generically.
```

Comment:
Is it a typo?

```suggestion
In _this_ situation `aws.by_region` in both the parent and child modules refers to the same multi-instance provider configuration, with the same instance keys. The parent module allows the child module to decide which instance keys to use, so the child module can be written generically.
```

Reply:
Reading this again, I think what I wrote is what I intended to write: my intent with this example is that the child module does not make any assumptions about what the instance keys will be. It just uses whatever instances the parent module passed it. So it's the parent module that is "choosing" the instance keys, and the child module is just reacting to that.